File: document/src/vespa/document/select/resultlist.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "resultlist.h"
#include <bitset>
#include <ostream>

namespace document::select {

ResultList::ResultList() = default;
ResultList::ResultList(ResultList &&) noexcept = default;
ResultList & ResultList::operator = (ResultList &&) noexcept = default;
ResultList::~ResultList() = default;

ResultList::ResultList(const Result& result) {
    add(VariableMap(), result);
}

void
ResultList::add(VariableMap variables, const Result& result)
{
    _results.emplace_back(std::move(variables), &result);
}

void
ResultList::print(std::ostream& out, bool, const std::string&) const
{
    out << "ResultList(";
    for (uint32_t i = 0; i < _results.size(); i++) {
        if (!_results[i].first.empty()) {
            out << _results[i].first.toString() << " => ";
        }
        out << _results[i].second->toString() << " ";
    }
    out << ")";
}

namespace {

const Result &
combineResultsLocal(const ResultList::Results & results)
{
    if (results.empty()) {
        return Result::False;
    }

    bool foundFalse = false;

    for (const auto & it : results) {
        if (*it.second == Result::True) {
            return Result::True;
        } else if (*it.second == Result::False) {
            foundFalse = true;
        }
    }

    return (foundFalse) ? Result::False : Result::Invalid;
}

}

const Result&
ResultList::combineResults() const {
    return combineResultsLocal(_results);
}

bool
ResultList::combineVariables(VariableMap& combination, const VariableMap& a, const VariableMap& b)
{
    // First, verify that all variables are overlapping
    for (const auto & ovar : a) {
        auto found(b.find(ovar.first));

        if (found != b.end()) {
            if (!(found->second == ovar.second)) {
                return false;
            }
        }
    }

    for (const auto & ivar : b) {
        auto found(a.find(ivar.first));
        if (found != a.end()) {
            if (!(found->second == ivar.second)) {
                return false;
            }
        }
    }
    // Ok, variables are overlapping. Add all variables from input to output.
    for (const auto & var : a) {
        combination[var.first] = var.second;
    }
    for (const auto & var : b) {
        combination[var.first] = var.second;
    }

    return true;
}

ResultList
ResultList::operator&&(const ResultList& other) const
{
    ResultList results;

    std::bitset<3> resultForNoVariables;
    for (const auto & it : _results) {
        for (const auto & it2 : other._results) {
            VariableMap vars;
            if ( combineVariables(vars, it.first, it2.first) ) {
                const Result & result = *it.second && *it2.second;
                if (vars.empty()) {
                    resultForNoVariables.set(result.toEnum());
                } else {
                    results.add(std::move(vars), result);
                }
            }
        }
    }
    for (uint32_t i(0); i < resultForNoVariables.size(); i++) {
        if (resultForNoVariables[i]) {
            results.add(VariableMap(), Result::fromEnum(i));
        }
    }

    return results;
}

ResultList
ResultList::operator||(const ResultList& other) const
{
    ResultList results;

    std::bitset<3> resultForNoVariables;
    for (const auto & it : _results) {
        for (const auto & it2 : other._results) {
            VariableMap vars;
            if (combineVariables(vars, it.first, it2.first)) {
                const Result & result = *it.second || *it2.second;
                if (vars.empty()) {
                    resultForNoVariables.set(result.toEnum());
                } else {
                    results.add(std::move(vars), result);
                }
            }
        }
    }
    for (uint32_t i(0); i < resultForNoVariables.size(); i++) {
        if (resultForNoVariables[i]) {
            results.add(VariableMap(), Result::fromEnum(i));
        }
    }

    return results;
}

ResultList
ResultList::operator!() && {
    ResultList result;

    for (auto & it : _results) {
        result.add(std::move(it.first), !*it.second);
    }

    return result;
}

bool ResultList::operator==(const ResultList& other) const {
    return (combineResultsLocal(_results) == combineResultsLocal(other._results));
}

}

File: document/src/main/java/com/yahoo/document/select/ResultList.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document.select;

import com.yahoo.document.datatypes.FieldPathIteratorHandler;
import com.yahoo.document.select.rule.AttributeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ResultList {
    public static class ResultPair {
        ResultPair(FieldPathIteratorHandler.VariableMap var, Result res) {
            variables = var;
            result = res;
        }

        FieldPathIteratorHandler.VariableMap variables;
        Result result;

        public FieldPathIteratorHandler.VariableMap getVariables() { return variables; }
        public Result getResult() { return result; }

        public String toString() {
            return variables.toString() + " => " + result;
        }
    }

    public static class VariableValue {
       public VariableValue(FieldPathIteratorHandler.VariableMap vars, Object value) {
           variables = vars;
           this.value = value;
       }

        FieldPathIteratorHandler.VariableMap variables;
        Object value;

        public FieldPathIteratorHandler.VariableMap getVariables() { return variables; }
        public Object getValue() { return value; }

        public String toString() {
            return variables.toString() + " => " + value;
        }
    }

    private List<ResultPair> results = new ArrayList<>();

    public ResultList() {
    }

    public ResultList(Result result) {
        add(new FieldPathIteratorHandler.VariableMap(), result);
    }

    public void add(FieldPathIteratorHandler.VariableMap variables, Result result) {
        results.add(new ResultPair(variables, result));
    }

    public List<ResultPair> getResults() {
        return results;
    }

    public static ResultList fromBoolean(boolean result) {
        return new ResultList(result ? Result.TRUE : Result.FALSE);
    }

    public Result toResult() {
        if (results.isEmpty()) {
            return Result.FALSE;
        }

        boolean foundFalse = false;

        for (ResultPair rp : results) {
            if (rp.result == Result.TRUE) {
                return Result.TRUE;
            } else if (rp.result == Result.FALSE) {
                foundFalse = true;
            }
        }

        if (foundFalse) {
            return Result.FALSE;
        } else {
            return Result.INVALID;
        }
    }

    private boolean combineVariables(FieldPathIteratorHandler.VariableMap output, FieldPathIteratorHandler.VariableMap input) {
        // First, verify that all variables are overlapping
        for (Map.Entry<String, FieldPathIteratorHandler.IndexValue> entry : output.entrySet()) {
            FieldPathIteratorHandler.IndexValue found = input.get(entry.getKey());

            if (found != null) {
                if (!(found.equals(entry.getValue()))) {
                    return false;
                }
            }
        }

        for (Map.Entry<String, FieldPathIteratorHandler.IndexValue> entry : input.entrySet()) {
            FieldPathIteratorHandler.IndexValue found = output.get(entry.getKey());

            if (found != null) {
                if (!(found.equals(entry.getValue()))) {
                    return false;
                }
            }
        }

        // Ok, variables are overlapping. Add all variables from input to output.
        for (Map.Entry<String, FieldPathIteratorHandler.IndexValue> entry : input.entrySet()) {
            output.put(entry.getKey(), entry.getValue());
        }

        return true;
    }

    public interface LazyResultList {
        ResultList getResult();
    }

    public ResultList combineAND(LazyResultList other)
    {
        if (Result.FALSE == toResult()) return ResultList.toResultList(false);

        ResultList result = new ResultList();

        // TODO: optimize
        for (ResultPair pair : results) {
            for (ResultPair otherPair : other.getResult().results) {
                FieldPathIteratorHandler.VariableMap varMap = (FieldPathIteratorHandler.VariableMap)pair.variables.clone();

                if (combineVariables(varMap, otherPair.variables)) {
                    result.add(varMap, combineAND(pair.result, otherPair.result));
                }
            }
        }

        return result;
    }

    private static Result combineAND(Result lhs, Result rhs) {
        if (lhs == Result.TRUE && rhs == Result.TRUE) {
            return Result.TRUE;
        }
        if (lhs == Result.FALSE || rhs == Result.FALSE) {
            return Result.FALSE;
        }
        return Result.INVALID;
    }

    public ResultList combineOR(LazyResultList other)
    {
        if (Result.TRUE == toResult()) return ResultList.toResultList(true);

        ResultList result = new ResultList();

        // TODO: optimize
        for (ResultPair pair : results) {
            for (ResultPair otherPair : other.getResult().results) {
                FieldPathIteratorHandler.VariableMap varMap = (FieldPathIteratorHandler.VariableMap)pair.variables.clone();

                if (combineVariables(varMap, otherPair.variables)) {
                    result.add(varMap, combineOR(pair.result, otherPair.result));
                }
            }
        }

        return result;
    }

    private static Result combineOR(Result lhs, Result rhs) {
        if (lhs == Result.TRUE || rhs == Result.TRUE) {
            return Result.TRUE;
        }
        if (lhs == Result.FALSE && rhs == Result.FALSE) {
            return Result.FALSE;
        }
        return Result.INVALID;
    }

    /**
     * Converts the given object value into a result list, so it can be compared by logical operators.
     *
     * @param value The object to convert.
     * @return The corresponding result value.
     */
    public static ResultList toResultList(Object value) {
        if (value instanceof ResultList) {
            return (ResultList)value;
        } else if (value instanceof AttributeNode.VariableValueList) {
            ResultList retVal = new ResultList();
            for (VariableValue vv : (AttributeNode.VariableValueList)value) {
                retVal.add(vv.getVariables(), Result.TRUE);
            }
            return retVal;
        } else if (value == null || value == Result.FALSE || value == Boolean.FALSE ||
            (value instanceof Number && ((Number)value).doubleValue() == 0)) {
            return new ResultList(Result.FALSE);
        } else if (value == Result.INVALID) {
            return new ResultList(Result.INVALID);
        } else {
            return new ResultList(Result.TRUE);
        }
    }

    public String toString() {
        return results.toString();
    }
}
